### DISCUSSION TITLE
Emgu.CV.runtime.windows.cuda 4.4.0.4099    CudaInvoke problem

### DISCUSSION THREAD

**Discussion by barabba2005** [emgucv/emgucv#655] [Q&A]
Hi all

I'm studing how use the GPU with Emgu.
I'm using a Nvidia Quadro K2000.
I tested the Cuda driver with CudaInvoke.HasCuda and the graphic Card work fine.

My program read a UDP Video stream:

using Emgu.CV.Cuda;
using Emgu.CV;
using Emgu.CV.Structure;
using Emgu.CV.CvEnum;

private VideoCapture videocapture = null;
private Image<Gray, byte> currenteFrame = null;
private bool faceDetectionEnable = false;
Mat Mframe = new Mat();
CascadeClassifier cascadeClassifier = new CascadeClassifier("haarcascade_frontalface_default.xml");

private void btnCapture_Click(object sender, EventArgs e)
        {
            videocapture = new Emgu.CV.VideoCapture("udp://239.255.255.255:2001", VideoCapture.API.Any);
            videocapture.ImageGrabbed += ProcessFrame;
            videocapture.Start();
        }

and try to make a face detector;

private void ProcessFrame(object sender, EventArgs e)
        {
            videocapture.Retrieve(Mframe, 0);
            currenteFrame = Mframe.ToImage<Gray, byte>().Resize(picCapture.Width, picCapture.Height, Inter.Linear);

            picCapture.Image = currenteFrame.ToBitmap<Gray, byte>();                          // pictureBox1 Video Input
            Thread.Sleep((int)Emgu.CV.CvEnum.CapProp.Fps);

            //Face Detection
            if (faceDetectionEnable)
            {
                //Convert to Gray Image
                Mat grayImm = new Mat();

                 //Enance image to get better result
                CvInvoke.EqualizeHist(currenteFrame, grayImm);           ​
               ​picAnalyzed.Image = grayImm.ToImage<Gray, byte>().ToBitmap();         // pictureBox2 Video Enhanced
}

All works fine if i use : 
Cvnvoke.EqualizeHist(currenteFrame, grayImm)
but if i try whit
CudaInvoke.EqualizeHist(currenteFrame, grayImm)
the program work, but video input stop and video Enhanced not appear.

Thank in advance for help.

**Comment by emgucv** [comment]
For CudaInvoke.EqualizeHist, both parameters should be CudaMat (instead of Mat).

Can you try to upload the currentFrame to a GpuMat, allocated a new GpuMat for the result, then download the result to grayImm?

**Comment by barabba2005** [comment]
Thank you for the replay,

I am moving forward in small steps.
If I write to
GpuMat grayImm = new GpuMat (); only,
the video input play fine, but if I add
grayImm.Upload (currenteFrame);
the video input stop.

I tried
GpuMat grayImm = new GpuMat ();
grayImm.Upload (currenteFrame);
CudaInvoke.EqualizeHist(currenteFrame, MyFrame);

but Video stop.

**Comment by barabba2005** [comment]
I tried like this:
GpuFrame.Upload(MatFrame); // Convert Mat to Gpu
GpuFrame.Download(MatGrayImm); // Convert Gpu to Mat

CvInvoke.EqualizeHist(currenteFrame, MatGrayImm);   // Apply EqualizeHist
picAnalyzed.Image = currenteFrame.ToBitmap<Bgr, byte>();

No work, when i add CvInvoke.EqualizeHist stop input image

Sorry, important question.
Where can I get the Emgu logs?
